certdb/certdb.go

```diff
@@ -0,0 +1,227 @@
+package certdb
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	cferr "github.com/cloudflare/cfssl/errors"
+	"github.com/kisielk/sqlstruct"
+)
+
+// CertificateRecord encodes a certificate and its metadata
+// that will be recorded in a database.
+type CertificateRecord struct {
+	Serial    string    `sql:"serial"`
+	CALabel   string    `sql:"ca_label"`
+	Status    string    `sql:"status"`
+	Reason    int       `sql:"reason"`
+	Expiry    time.Time `sql:"expiry"`
+	RevokedAt time.Time `sql:"revoked_at"`
+	PEM       string    `sql:"pem"`
+}
+
+// OCSPRecord encodes a OCSP response body and its metadata
+// that will be recorded in a database.
+type OCSPRecord struct {
+	Serial string    `sql:"serial"`
+	Body   string    `sql:"body"`
+	Expiry time.Time `sql:"expiry"`
+}
+
+const (
+	insertSQL = `
+INSERT INTO certificates (serial, ca_label, status, reason, expiry, revoked_at, pem)
+	VALUES ($1, $2, $3, $4, $5, $6, $7);`
+
+	selectSQL = `
+SELECT %s FROM certificates
+	WHERE (serial = $1);`
+
+	selectAllSQL = `
+SELECT %s FROM certificates;`
+
+	selectAllUnexpiredSQL = `
+SELECT %s FROM certificates
+WHERE CURRENT_TIMESTAMP < expiry;`
+
+	updateRevokeSQL = `
+UPDATE certificates
+	SET status='revoked', revoked_at=CURRENT_TIMESTAMP, reason=$1
+	WHERE (serial = $2);`
+
+	insertOCSPSQL = `
+INSERT INTO ocsp_responses (serial, body, expiry)
+    VALUES ($1, $2, $3);`
+
+	updateOCSPSQL = `
+UPDATE ocsp_responses
+    SET expiry=$3, body=$2
+	WHERE (serial = $1);`
+
+	selectAllUnexpiredOCSPSQL = `
+SELECT %s FROM ocsp_responses
+WHERE CURRENT_TIMESTAMP < expiry;`
+
+	selectOCSPSQL = `
+SELECT %s FROM ocsp_responses
+    WHERE (serial = $1);`
+)
+
+func wrapCertStoreError(err error) error {
+	if err != nil {
+		return cferr.Wrap(cferr.CertStoreError, cferr.Unknown, err)
+	}
+	return nil
+}
+
+// InsertCertificate puts a CertificateRecord into db.
+func InsertCertificate(db *sql.DB, cr *CertificateRecord) error {
+	res, err := db.Exec(
+		insertSQL,
+		cr.Serial,
+		cr.CALabel,
+		cr.Status,
+		cr.Reason,
+		cr.Expiry.UTC(),
```

Comment:
I think it is fine. https://golang.org/src/time/time.go?s=2427:2459#L39. Time is kept as the seconds and nseconds since the epoch. Since in sqlite we use 'timestamp' (without timezone), we convert them into UTC, and make them consistent to be compared with CURRENT_TIMESTAMP (a UTC timestamp). Time conversion can be easily done by changing Location, but the underlying representation has not changed.